### DISCUSSION TITLE
NotificationCenter Not Triggering for Background App

### DISCUSSION THREAD

**Discussion by chigkim** [sparkle-project/Sparkle#2492] [Q&A] [answered]
My ap runs in the background on the menu extra. If I check the update manually, it works, but I don't get any notification center alert.
I also set NSUserNotificationAlertStyle to alert, and the app is notarized.
After running the app, Notification center asked me to allow my app to send notification. I allowed it.
My implementation is below.
I'd appreciate any help!
Thanks!

```
import Cocoa
import Sparkle
import UserNotifications

class AutoUpdateManager: NSObject, SPUUpdaterDelegate, SPUStandardUserDriverDelegate, UNUserNotificationCenterDelegate {

	static let shared = AutoUpdateManager()
	private var updaterController: SPUStandardUpdaterController?
	private let UPDATE_NOTIFICATION_IDENTIFIER = "VOCRUpdateCheck"
		var supportsGentleScheduledUpdateReminders: Bool {
		return true
	}

	private override init() {
		super.init()
		self.updaterController = SPUStandardUpdaterController(startingUpdater: true, updaterDelegate: self, userDriverDelegate: self)
		UNUserNotificationCenter.current().delegate = self
		self.setupAutoUpdate()
	}

	private func setupAutoUpdate() {
		updaterController?.updater.automaticallyChecksForUpdates = true
		updaterController?.updater.automaticallyDownloadsUpdates = true
		updaterController?.updater.checkForUpdatesInBackground()
		updaterController?.updater.updateCheckInterval = 3600  // Check every hour
	}

	func checkForUpdates() {
		updaterController?.checkForUpdates(nil)
	}

	func updater(_ updater: SPUUpdater, willScheduleUpdateCheckAfterDelay delay: TimeInterval) {
		UNUserNotificationCenter.current().requestAuthorization(options: [.badge, .alert, .sound]) { granted, error in
			// Examine granted outcome and error if desired...
		}
	}
	
	func standardUserDriverWillHandleShowingUpdate(_ handleShowingUpdate: Bool, forUpdate update: SUAppcastItem, state: SPUUserUpdateState) {
		NSApp.setActivationPolicy(.regular)
		if !state.userInitiated {
			NSApp.dockTile.badgeLabel = "1"
			do {
				let content = UNMutableNotificationContent()
				content.title = "A new update is available"
				content.body = "Version \(update.displayVersionString) is now available"
				content.sound = UNNotificationSound.default
				let request = UNNotificationRequest(identifier: UPDATE_NOTIFICATION_IDENTIFIER, content: content, trigger: nil)
				UNUserNotificationCenter.current().add(request)
			}
		}
	}

	func standardUserDriverDidReceiveUserAttention(forUpdate update: SUAppcastItem) {
		NSApp.dockTile.badgeLabel = ""
		UNUserNotificationCenter.current().removeDeliveredNotifications(withIdentifiers: [UPDATE_NOTIFICATION_IDENTIFIER])
	}
	
	func standardUserDriverWillFinishUpdateSession() {
		NSApp.setActivationPolicy(.accessory)
	}
	
	func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
		if response.notification.request.identifier == UPDATE_NOTIFICATION_IDENTIFIER && response.actionIdentifier == UNNotificationDefaultActionIdentifier {
			updaterController?.checkForUpdates(nil)
		}
		completionHandler()
	}

}
```

**Comment by zorgiepoo** [comment, marked as answer]
Quick few things:

* Remove your setupAutoUpdate method. It's all wrong. Read the API documentation on these methods and/or the API Expecations section of the [programmatic setup page](https://sparkle-project.org/documentation/programmatic-setup/#api-expectations). Hopefully you do not use APIs without reading their documentation.
* A notification only shows up by default when the app is in the background so you need to test an update being delivered when the app is not frontmost. This example also removes the delivered notification when the user notices the update. Near app launch, Sparkle may show the update in a frontmost state which would have no need for a notification. To test when an update is delivered and Sparkle's update window does not make the app the active/frontmost one, look at the second defaults write example in the [gentle reminders documentation](https://sparkle-project.org/documentation/gentle-reminders/) before launching an app (requires removing the checkForUpdatesInBackground check in prior point).
* If Do Not Disturb is on, notification may not deliver. Also the user needs to approve notifications but you already confirmed that.
* If automatic downloading of updates is enabled (which you set at runtime and now persists in the user defaults of the app until you clear it), an update may not be presented to the user at all (in a short timely manner) and may be installed on quit. This is a "silently install update" option where if enabled an update alert may only be automatically shown one week later if the app hasn't been quit yet. If the user needs to authorize/enter their admin password to install an update or the update is marked as critical/informational/major upgrade, Sparkle will schedule an update alert more immediately however.

[edit]: to clarify automatic downloading option too.

**Reply by chigkim** [reply]
thanks for your response.

Could you explain what do you mean by "It's all wrong?"

I have setup menu to check update according to the [Setting up Sparkle programmatically](https://sparkle-project.org/documentation/programmatic-setup/) you pointed out, and that works correctly.

Also my code is pretty much copied from [gentle reminders documentation](https://sparkle-project.org/documentation/gentle-reminders/).

**Reply by zorgiepoo** [reply]
Your `setupAutoUpdate` method which contains code you did not copy from the documentation. Re-read what I said. Read the API expectations section and/or the API documentation of all the APIs you are calling in that method.

**Reply by chigkim** [reply]
Ah ok that makes sense. Basically I shouldn't mess with those settings and leave them up to the users.

If I want to at least check for update when app is starting, can I use checkForUpdatesInBackground, or do you recommend not to use that either?

Thanks for your help!

**Reply by zorgiepoo** [reply]
> If I want to at least check for update when app is starting, can I use checkForUpdatesInBackground, or do you recommend not to use that either?

Yeah that is fine if you want to do that. If you do a checkForUpdatesInBackground at app launch, set [SUEnableAutomaticChecks](https://sparkle-project.org/documentation/customization/) in your Info.plist so Sparkle won't think it needs to trigger a permission prompt on the app's second launch (asking the user if they want sparkle to check for updates automatically). Or only call `checkForUpdatesInBackground` when `updater.automaticallyChecksForUpdates` is `YES` if you still want Sparkle to prompt asking the user for permission to check for updates automatically (and then you won't trigger an update check before the user approved of automatic update checking..)

Back to the original issue if you want it to be easier to test Sparkle bringing up an update alert in the background and you delivering a user notification, at least for testing you may not want to opt into automatic downloading/installing of updates. Also always performing a background update check near launch will also make it more difficult to test this as Sparkle has a preference of showing updates up-front near launch. However you can still wait for the next scheduled time interval (which is 1 day by default, but you were setting it to 1 hour).

**Comment by chigkim** [comment]
For people who ran alpha builds before, should I just delete SUEnableAutomaticChecks, SUScheduledCheckInterval, SUAutomaticallyUpdate from their UserDefaults?

**Reply by zorgiepoo** [reply]
It's up to you I guess. Not sure how pretty it will be and how much they diverge from the defaults you still want to provide in the Info.plist. Check the [customization page](https://sparkle-project.org/documentation/customization/). You can adjust the initial default settings there, although user default settings will override the Info.plist ones, which is why the programmatic APIs for these should only be done through the user changing settings.

**Comment by chigkim** [comment]
Thank you so much! You've been very helpful!
